// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: api/feedback/v1/feedback.proto

package v1

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on Feedback with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *Feedback) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on Feedback with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in FeedbackMultiError, or nil
// if none found.
func (m *Feedback) ValidateAll() error {
	return m.validate(true)
}

func (m *Feedback) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	// no validation rules for Name

	// no validation rules for Content

	// no validation rules for Like

	if len(errors) > 0 {
		return FeedbackMultiError(errors)
	}

	return nil
}

// FeedbackMultiError is an error wrapping multiple validation errors returned
// by Feedback.ValidateAll() if the designated constraints aren't met.
type FeedbackMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m FeedbackMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m FeedbackMultiError) AllErrors() []error { return m }

// FeedbackValidationError is the validation error returned by
// Feedback.Validate if the designated constraints aren't met.
type FeedbackValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e FeedbackValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e FeedbackValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e FeedbackValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e FeedbackValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e FeedbackValidationError) ErrorName() string { return "FeedbackValidationError" }

// Error satisfies the builtin error interface
func (e FeedbackValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sFeedback.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = FeedbackValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = FeedbackValidationError{}

// Validate checks the field values on CreateFeedbackRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *CreateFeedbackRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CreateFeedbackRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// CreateFeedbackRequestMultiError, or nil if none found.
func (m *CreateFeedbackRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *CreateFeedbackRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if l := utf8.RuneCountInString(m.GetName()); l < 5 || l > 50 {
		err := CreateFeedbackRequestValidationError{
			field:  "Name",
			reason: "value length must be between 5 and 50 runes, inclusive",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	// no validation rules for Content

	if len(errors) > 0 {
		return CreateFeedbackRequestMultiError(errors)
	}

	return nil
}

// CreateFeedbackRequestMultiError is an error wrapping multiple validation
// errors returned by CreateFeedbackRequest.ValidateAll() if the designated
// constraints aren't met.
type CreateFeedbackRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CreateFeedbackRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CreateFeedbackRequestMultiError) AllErrors() []error { return m }

// CreateFeedbackRequestValidationError is the validation error returned by
// CreateFeedbackRequest.Validate if the designated constraints aren't met.
type CreateFeedbackRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CreateFeedbackRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CreateFeedbackRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CreateFeedbackRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CreateFeedbackRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CreateFeedbackRequestValidationError) ErrorName() string {
	return "CreateFeedbackRequestValidationError"
}

// Error satisfies the builtin error interface
func (e CreateFeedbackRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCreateFeedbackRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CreateFeedbackRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CreateFeedbackRequestValidationError{}

// Validate checks the field values on CreateFeedbackReply with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *CreateFeedbackReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CreateFeedbackReply with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// CreateFeedbackReplyMultiError, or nil if none found.
func (m *CreateFeedbackReply) ValidateAll() error {
	return m.validate(true)
}

func (m *CreateFeedbackReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetFeedback()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, CreateFeedbackReplyValidationError{
					field:  "Feedback",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, CreateFeedbackReplyValidationError{
					field:  "Feedback",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetFeedback()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return CreateFeedbackReplyValidationError{
				field:  "Feedback",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return CreateFeedbackReplyMultiError(errors)
	}

	return nil
}

// CreateFeedbackReplyMultiError is an error wrapping multiple validation
// errors returned by CreateFeedbackReply.ValidateAll() if the designated
// constraints aren't met.
type CreateFeedbackReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CreateFeedbackReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CreateFeedbackReplyMultiError) AllErrors() []error { return m }

// CreateFeedbackReplyValidationError is the validation error returned by
// CreateFeedbackReply.Validate if the designated constraints aren't met.
type CreateFeedbackReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CreateFeedbackReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CreateFeedbackReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CreateFeedbackReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CreateFeedbackReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CreateFeedbackReplyValidationError) ErrorName() string {
	return "CreateFeedbackReplyValidationError"
}

// Error satisfies the builtin error interface
func (e CreateFeedbackReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCreateFeedbackReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CreateFeedbackReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CreateFeedbackReplyValidationError{}

// Validate checks the field values on UpdateFeedbackRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *UpdateFeedbackRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UpdateFeedbackRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// UpdateFeedbackRequestMultiError, or nil if none found.
func (m *UpdateFeedbackRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *UpdateFeedbackRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetId() <= 0 {
		err := UpdateFeedbackRequestValidationError{
			field:  "Id",
			reason: "value must be greater than 0",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if l := utf8.RuneCountInString(m.GetName()); l < 5 || l > 50 {
		err := UpdateFeedbackRequestValidationError{
			field:  "Name",
			reason: "value length must be between 5 and 50 runes, inclusive",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	// no validation rules for Content

	if len(errors) > 0 {
		return UpdateFeedbackRequestMultiError(errors)
	}

	return nil
}

// UpdateFeedbackRequestMultiError is an error wrapping multiple validation
// errors returned by UpdateFeedbackRequest.ValidateAll() if the designated
// constraints aren't met.
type UpdateFeedbackRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UpdateFeedbackRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UpdateFeedbackRequestMultiError) AllErrors() []error { return m }

// UpdateFeedbackRequestValidationError is the validation error returned by
// UpdateFeedbackRequest.Validate if the designated constraints aren't met.
type UpdateFeedbackRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UpdateFeedbackRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UpdateFeedbackRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UpdateFeedbackRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UpdateFeedbackRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UpdateFeedbackRequestValidationError) ErrorName() string {
	return "UpdateFeedbackRequestValidationError"
}

// Error satisfies the builtin error interface
func (e UpdateFeedbackRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUpdateFeedbackRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UpdateFeedbackRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UpdateFeedbackRequestValidationError{}

// Validate checks the field values on UpdateFeedbackReply with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *UpdateFeedbackReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UpdateFeedbackReply with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// UpdateFeedbackReplyMultiError, or nil if none found.
func (m *UpdateFeedbackReply) ValidateAll() error {
	return m.validate(true)
}

func (m *UpdateFeedbackReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetFeedback()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, UpdateFeedbackReplyValidationError{
					field:  "Feedback",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, UpdateFeedbackReplyValidationError{
					field:  "Feedback",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetFeedback()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return UpdateFeedbackReplyValidationError{
				field:  "Feedback",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return UpdateFeedbackReplyMultiError(errors)
	}

	return nil
}

// UpdateFeedbackReplyMultiError is an error wrapping multiple validation
// errors returned by UpdateFeedbackReply.ValidateAll() if the designated
// constraints aren't met.
type UpdateFeedbackReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UpdateFeedbackReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UpdateFeedbackReplyMultiError) AllErrors() []error { return m }

// UpdateFeedbackReplyValidationError is the validation error returned by
// UpdateFeedbackReply.Validate if the designated constraints aren't met.
type UpdateFeedbackReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UpdateFeedbackReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UpdateFeedbackReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UpdateFeedbackReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UpdateFeedbackReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UpdateFeedbackReplyValidationError) ErrorName() string {
	return "UpdateFeedbackReplyValidationError"
}

// Error satisfies the builtin error interface
func (e UpdateFeedbackReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUpdateFeedbackReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UpdateFeedbackReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UpdateFeedbackReplyValidationError{}

// Validate checks the field values on DeleteFeedbackRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *DeleteFeedbackRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DeleteFeedbackRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// DeleteFeedbackRequestMultiError, or nil if none found.
func (m *DeleteFeedbackRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *DeleteFeedbackRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return DeleteFeedbackRequestMultiError(errors)
	}

	return nil
}

// DeleteFeedbackRequestMultiError is an error wrapping multiple validation
// errors returned by DeleteFeedbackRequest.ValidateAll() if the designated
// constraints aren't met.
type DeleteFeedbackRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DeleteFeedbackRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DeleteFeedbackRequestMultiError) AllErrors() []error { return m }

// DeleteFeedbackRequestValidationError is the validation error returned by
// DeleteFeedbackRequest.Validate if the designated constraints aren't met.
type DeleteFeedbackRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DeleteFeedbackRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DeleteFeedbackRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DeleteFeedbackRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DeleteFeedbackRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DeleteFeedbackRequestValidationError) ErrorName() string {
	return "DeleteFeedbackRequestValidationError"
}

// Error satisfies the builtin error interface
func (e DeleteFeedbackRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDeleteFeedbackRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DeleteFeedbackRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DeleteFeedbackRequestValidationError{}

// Validate checks the field values on DeleteFeedbackReply with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *DeleteFeedbackReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DeleteFeedbackReply with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// DeleteFeedbackReplyMultiError, or nil if none found.
func (m *DeleteFeedbackReply) ValidateAll() error {
	return m.validate(true)
}

func (m *DeleteFeedbackReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return DeleteFeedbackReplyMultiError(errors)
	}

	return nil
}

// DeleteFeedbackReplyMultiError is an error wrapping multiple validation
// errors returned by DeleteFeedbackReply.ValidateAll() if the designated
// constraints aren't met.
type DeleteFeedbackReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DeleteFeedbackReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DeleteFeedbackReplyMultiError) AllErrors() []error { return m }

// DeleteFeedbackReplyValidationError is the validation error returned by
// DeleteFeedbackReply.Validate if the designated constraints aren't met.
type DeleteFeedbackReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DeleteFeedbackReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DeleteFeedbackReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DeleteFeedbackReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DeleteFeedbackReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DeleteFeedbackReplyValidationError) ErrorName() string {
	return "DeleteFeedbackReplyValidationError"
}

// Error satisfies the builtin error interface
func (e DeleteFeedbackReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDeleteFeedbackReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DeleteFeedbackReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DeleteFeedbackReplyValidationError{}

// Validate checks the field values on GetFeedbackRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *GetFeedbackRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetFeedbackRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetFeedbackRequestMultiError, or nil if none found.
func (m *GetFeedbackRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *GetFeedbackRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return GetFeedbackRequestMultiError(errors)
	}

	return nil
}

// GetFeedbackRequestMultiError is an error wrapping multiple validation errors
// returned by GetFeedbackRequest.ValidateAll() if the designated constraints
// aren't met.
type GetFeedbackRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetFeedbackRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetFeedbackRequestMultiError) AllErrors() []error { return m }

// GetFeedbackRequestValidationError is the validation error returned by
// GetFeedbackRequest.Validate if the designated constraints aren't met.
type GetFeedbackRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetFeedbackRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetFeedbackRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetFeedbackRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetFeedbackRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetFeedbackRequestValidationError) ErrorName() string {
	return "GetFeedbackRequestValidationError"
}

// Error satisfies the builtin error interface
func (e GetFeedbackRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetFeedbackRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetFeedbackRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetFeedbackRequestValidationError{}

// Validate checks the field values on GetFeedbackReply with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *GetFeedbackReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetFeedbackReply with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetFeedbackReplyMultiError, or nil if none found.
func (m *GetFeedbackReply) ValidateAll() error {
	return m.validate(true)
}

func (m *GetFeedbackReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetFeedback()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, GetFeedbackReplyValidationError{
					field:  "Feedback",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, GetFeedbackReplyValidationError{
					field:  "Feedback",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetFeedback()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return GetFeedbackReplyValidationError{
				field:  "Feedback",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return GetFeedbackReplyMultiError(errors)
	}

	return nil
}

// GetFeedbackReplyMultiError is an error wrapping multiple validation errors
// returned by GetFeedbackReply.ValidateAll() if the designated constraints
// aren't met.
type GetFeedbackReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetFeedbackReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetFeedbackReplyMultiError) AllErrors() []error { return m }

// GetFeedbackReplyValidationError is the validation error returned by
// GetFeedbackReply.Validate if the designated constraints aren't met.
type GetFeedbackReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetFeedbackReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetFeedbackReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetFeedbackReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetFeedbackReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetFeedbackReplyValidationError) ErrorName() string { return "GetFeedbackReplyValidationError" }

// Error satisfies the builtin error interface
func (e GetFeedbackReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetFeedbackReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetFeedbackReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetFeedbackReplyValidationError{}

// Validate checks the field values on ListFeedbackRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ListFeedbackRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListFeedbackRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListFeedbackRequestMultiError, or nil if none found.
func (m *ListFeedbackRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ListFeedbackRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return ListFeedbackRequestMultiError(errors)
	}

	return nil
}

// ListFeedbackRequestMultiError is an error wrapping multiple validation
// errors returned by ListFeedbackRequest.ValidateAll() if the designated
// constraints aren't met.
type ListFeedbackRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListFeedbackRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListFeedbackRequestMultiError) AllErrors() []error { return m }

// ListFeedbackRequestValidationError is the validation error returned by
// ListFeedbackRequest.Validate if the designated constraints aren't met.
type ListFeedbackRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListFeedbackRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListFeedbackRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListFeedbackRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListFeedbackRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListFeedbackRequestValidationError) ErrorName() string {
	return "ListFeedbackRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ListFeedbackRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListFeedbackRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListFeedbackRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListFeedbackRequestValidationError{}

// Validate checks the field values on ListFeedbackReply with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *ListFeedbackReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListFeedbackReply with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListFeedbackReplyMultiError, or nil if none found.
func (m *ListFeedbackReply) ValidateAll() error {
	return m.validate(true)
}

func (m *ListFeedbackReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetResults() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ListFeedbackReplyValidationError{
						field:  fmt.Sprintf("Results[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ListFeedbackReplyValidationError{
						field:  fmt.Sprintf("Results[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ListFeedbackReplyValidationError{
					field:  fmt.Sprintf("Results[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return ListFeedbackReplyMultiError(errors)
	}

	return nil
}

// ListFeedbackReplyMultiError is an error wrapping multiple validation errors
// returned by ListFeedbackReply.ValidateAll() if the designated constraints
// aren't met.
type ListFeedbackReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListFeedbackReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListFeedbackReplyMultiError) AllErrors() []error { return m }

// ListFeedbackReplyValidationError is the validation error returned by
// ListFeedbackReply.Validate if the designated constraints aren't met.
type ListFeedbackReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListFeedbackReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListFeedbackReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListFeedbackReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListFeedbackReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListFeedbackReplyValidationError) ErrorName() string {
	return "ListFeedbackReplyValidationError"
}

// Error satisfies the builtin error interface
func (e ListFeedbackReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListFeedbackReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListFeedbackReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListFeedbackReplyValidationError{}

// Validate checks the field values on LikeRequest with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *LikeRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on LikeRequest with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in LikeRequestMultiError, or
// nil if none found.
func (m *LikeRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *LikeRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	// no validation rules for Unlike

	if len(errors) > 0 {
		return LikeRequestMultiError(errors)
	}

	return nil
}

// LikeRequestMultiError is an error wrapping multiple validation errors
// returned by LikeRequest.ValidateAll() if the designated constraints aren't met.
type LikeRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m LikeRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m LikeRequestMultiError) AllErrors() []error { return m }

// LikeRequestValidationError is the validation error returned by
// LikeRequest.Validate if the designated constraints aren't met.
type LikeRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e LikeRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e LikeRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e LikeRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e LikeRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e LikeRequestValidationError) ErrorName() string { return "LikeRequestValidationError" }

// Error satisfies the builtin error interface
func (e LikeRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sLikeRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = LikeRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = LikeRequestValidationError{}

// Validate checks the field values on LikeReply with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *LikeReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on LikeReply with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in LikeReplyMultiError, or nil
// if none found.
func (m *LikeReply) ValidateAll() error {
	return m.validate(true)
}

func (m *LikeReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return LikeReplyMultiError(errors)
	}

	return nil
}

// LikeReplyMultiError is an error wrapping multiple validation errors returned
// by LikeReply.ValidateAll() if the designated constraints aren't met.
type LikeReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m LikeReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m LikeReplyMultiError) AllErrors() []error { return m }

// LikeReplyValidationError is the validation error returned by
// LikeReply.Validate if the designated constraints aren't met.
type LikeReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e LikeReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e LikeReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e LikeReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e LikeReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e LikeReplyValidationError) ErrorName() string { return "LikeReplyValidationError" }

// Error satisfies the builtin error interface
func (e LikeReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sLikeReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = LikeReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = LikeReplyValidationError{}
